homeassistant/components/orange_funbox/device_tracker.py

```diff
@@ -0,0 +1,146 @@
+"""Support for Orange FunBox router."""
+from collections import namedtuple
+import logging
+from typing import Any, Dict, List, Optional
+
+import requests
+import voluptuous as vol
+
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.typing import (
+    HomeAssistantType,
+    ConfigType,
+)
+import homeassistant.util.dt as dt_util
+from homeassistant.components.device_tracker import (
+    DOMAIN,
+    PLATFORM_SCHEMA,
+    DeviceScanner,
+)
+from homeassistant.const import CONF_HOST
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_EXCLUDE = 'exclude'
+
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_HOST): cv.string,
+    vol.Optional(CONF_EXCLUDE, default=[]):
+        vol.All(cv.ensure_list, [cv.string]),
+})
+
+
+Device = namedtuple(
+    'Device',
+    [
+        'active',
+        'device_type',
+        'first_seen',
+        'ip',
+        'last_connection',
+        'last_update',
+        'mac',
+        'name',
+        'signal_noise_ratio',
+        'signal_strength',
+    ],
+)
+
+
+class FunboxDeviceScanner(DeviceScanner):
+    """Device tracker for Orange FunBox 3."""
+
+    def __init__(self, config: ConfigType) -> None:
+        """Init."""
+        self.last_results = []
+
+        self.host = config[CONF_HOST]
+        self.exclude = config[CONF_EXCLUDE]
+
+        _LOGGER.debug('Scanner initialized')
+
+    def scan_devices(self) -> List[str]:
+        """Update list of devices from FunBox API."""
+        self._update_info()
+        return [
+            device.mac
+            for device in self.last_results
+        ]
+
+    def get_device_name(self, device: str) -> Optional[str]:
+        """Return the name of the given device or None."""
+        result = next(
+            (
+                result for result in self.last_results
+                if result.mac == device
+            ),
+            None,
+        )
+        return result.name if result else None
+
+    def get_extra_attributes(self, device: str) -> Optional[Dict[str, Any]]:
+        """Return all the attributes of the given device."""
+        result = next(
+            (
+                result for result in self.last_results
+                if result.mac == device
+            ),
+            None,
+        )
+        return result._asdict() if result else None
+
+    def _update_info(self) -> bool:
+        """Request list of devices from FunBox API."""
+        _LOGGER.debug('Getting devices')
+
+        path = '/sysbus/Devices:get'
+        resp = requests.post('http://{0}{1}'.format(self.host, path))
```

Comment:
Please extract the device specific interface code, I/O and data parsing, to a standalone library published on pypi.

https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#4-communication-with-devices-services